cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -727,6 +729,178 @@ struct CastFunctor<T, DictionaryType,
   }
 };
 
+// ----------------------------------------------------------------------
+// String to Number
+
+// Cast a string to a number.  Returns true on success, false on error.
+// We rely on C++ istringstream for locale-independent parsing, which might
+// not be the fastest option.
+
+template <typename T>
+typename std::enable_if<std::is_floating_point<T>::value,
+                        bool>::type static CastStringToNumber(std::istringstream& ibuf,
+                                                              T* out) {
+  ibuf >> *out;
+  return !ibuf.fail() && ibuf.eof();
+}
+
+// For integers, not all integer widths are handled by the C++ stdlib, so
+// we check for limits outselves.
+
+template <typename T>
+typename std::enable_if<std::is_integral<T>::value && std::is_signed<T>::value,
+                        bool>::type static CastStringToNumber(std::istringstream& ibuf,
+                                                              T* out) {
+  static constexpr bool need_long_long = sizeof(T) > sizeof(long);  // NOLINT
+  static constexpr T min_value = std::numeric_limits<T>::min();
+  static constexpr T max_value = std::numeric_limits<T>::max();
+
+  if (need_long_long) {
+    long long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res < min_value || res > max_value) {
+      return false;
+    }
+  } else {
+    long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res < min_value || res > max_value) {
+      return false;
+    }
+  }
+  return !ibuf.fail() && ibuf.eof();
+}
+
+template <typename T>
+typename std::enable_if<std::is_integral<T>::value && std::is_unsigned<T>::value,
+                        bool>::type static CastStringToNumber(std::istringstream& ibuf,
+                                                              T* out) {
+  static constexpr bool need_long_long = sizeof(T) > sizeof(unsigned long);  // NOLINT
+  static constexpr T max_value = std::numeric_limits<T>::max();
+
+  if (need_long_long) {
+    unsigned long long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res > max_value) {
+      return false;
+    }
+  } else {
+    unsigned long res;  // NOLINT
+    ibuf >> res;
+    *out = static_cast<T>(res);  // may downcast
+    if (res > max_value) {
+      return false;
+    }
+  }
+  return !ibuf.fail() && ibuf.eof();
+}
+
+template <typename O>
+struct CastFunctor<O, StringType, typename std::enable_if<is_number<O>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using out_type = typename O::c_type;
+
+    StringArray input_array(input.Copy());
+    auto out_data = GetMutableValues<out_type>(output, 1);
+    errno = 0;
+    // Instantiate the stringstream outside of the loop
+    std::istringstream ibuf;
+    ibuf.imbue(std::locale::classic());
```

Comment:
I sometimes wonder what the meeting was like that decided names like this.

Reply:
I had the same thought :-)